app/services/determine_training_record_state.rb

```diff
@@ -148,39 +148,23 @@ class DetermineTrainingRecordState < BaseService
     not_registered_for_training
   ].each_with_object({}) { |v, h| h[v] = v }.freeze
 
-  Record = Struct.new(
-    :participant_profile_id,
-    :induction_record_id,
-    :school_id,
-    :lead_provider_id,
-    :delivery_partner_id,
-    :appropriate_body_id,
-    :changed_at,
-    :validation_state,
-    :training_eligibility_state,
-    :fip_funding_eligibility_state,
-    :mentoring_state,
-    :training_state,
-    :record_state,
-    keyword_init: true,
-  ) do
-    def validation_status_valid?
-      validation_state == "valid"
-    end
-  end
-
   def call
-    result = ActiveRecord::Base.connection.execute(query)
+    if (training_record_state = TrainingRecordState.find_by(**params))
+      training_record_state.assign_attributes(**execute_query.first)
 
-    Record.new(**result.first)
+      training_record_state.save! if training_record_state.changed?
+      training_record_state
+    else
+      TrainingRecordState.create!(**execute_query.first)
+    end
```

Comment:
This is the slightly complicated bit I need to revisit and add plenty of tests around.